Поправлена валидация

Liza1040 opened the pull request.

EuphoriaAbsorber committed:
changed url for getproducts

EuphoriaAbsorber committed:
changed path for get products

Liza1040 committed:
Добавлена проверка на пустое поле и отрисовка ошибок после нажатия кнопки

Liza1040 committed:
Поправлена валидация на странице входа, добавлен обработчик события focusin

Liza1040 committed:
Создан класс для отрисовки ошибок и вынесен в отдельный модуль

Liza1040 committed:
Поправлена валидация на странице регистрации, добавлен обработчик события focusin

tUnknownLegend commented on public/src/pages/LoginPage/LoginPage.js:
```diff
@@ -45,88 +46,110 @@ export default class LoginPage extends BasePage {
          */
         const onSubmitHandler = async (event) => {
-            const data = [];
             const validation = new Val();
```
'validation' is assigned a value but never used

tUnknownLegend commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,121 @@ export default class RegisterPage extends BasePage {
+     * @param {object} event событие фокусирования на элементе
+     */
+     async DeleteErrorMessage(event) {
+        if (document.getElementById(event.target.name + 'Error') !== null) {
```
```suggestion
        if (document.getElementById(event.target.name + 'Error')) {
```

tUnknownLegend commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,121 @@ export default class RegisterPage extends BasePage {
+     * @param {object} data - объект, содержащий данные из формы
+     * @return {boolean} статус валидации
+     */
+     validate(data) {
```
Эта функция у меня в ПР поуниверсальней перписана. Но как-нибудь совместим ¯\_(ツ)_/¯

tUnknownLegend commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,121 @@ export default class RegisterPage extends BasePage {
+                form.removeEventListener('submit', onSubmitHandler);
+                config.header.main.render(config);
+                break;
+            case 400:
```
Стоит подумать о том, чтобы вынести сообщения об ошибке для каждого кода в модуль ErrorMessages. Но это так, закидушка в будущее, может что получше придумаем

DPeshkoff replied:
Сами сообщения лучше вообще вынести в константы, чтобы потом, например, было легче переводить

tUnknownLegend commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,121 @@ export default class RegisterPage extends BasePage {
+     * Метод, удаляющий сообщение об ошибке при фокусе на поле ввода
+     * @param {object} event событие фокусирования на элементе
+     */
+     async DeleteErrorMessage(event) {
```
лучше сделать отдельным методом, мб даже вынести в модуль ErrorMessages в каком-то универсальном формате

Liza1040 replied:
Вынесла в модуль ErrorMessage

tUnknownLegend commented on public/src/pages/LoginPage/LoginPage.js:
```diff
@@ -45,88 +46,110 @@ export default class LoginPage extends BasePage {
+     * @param {object} data - объект, содержащий данные из формы
+     * @return {boolean} статус валидации
+     */
+    validate(data) {
```
это вынесем в модуль ErrorMessages потом

Liza1040 replied:
Эту функцию я бы тут оставила) Ну, думаю, мы это еще обсудим)

DPeshkoff replied:
Обертка над модулем ок, только создавать инстансы, как мне кажется, лишнее
Лучше сделать глобальные синглтоны под эту роль

tUnknownLegend commented on public/src/pages/LoginPage/LoginPage.js:
```diff
@@ -45,88 +46,110 @@ export default class LoginPage extends BasePage {
+     * Метод, удаляющий сообщение об ошибке при фокусе на поле ввода
+     * @param {object} event событие фокусирования на элементе
+     */
+    async DeleteErrorMessage(event) {
```
по аналогии с регистрацией. Это всего логина касается

tUnknownLegend commented:
В целом замечания скорее по тому, что надо будет сделать в будущем, а не то, что должно быть готово уже сейчас

tUnknownLegend commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,121 @@ export default class RegisterPage extends BasePage {
+     */
+     async DeleteErrorMessage(event) {
+        if (document.getElementById(event.target.name + 'Error') !== null) {
+            document.getElementById(event.target.name + 'Error').remove();
```
Два раза искать одно и тоже два раза может быть не лучшая идея

Liza1040 committed:
Функция DeleteErrorMessage вынесена в модуль ErrorMessage

Liza1040 committed:
Добавлен async на обработчики события  focusin

tUnknownLegend commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,121 @@ export default class RegisterPage extends BasePage {
+            }
+
+            const r = new Req();
+            const [username, email, password, anotherPassword] = Array.from(data);
```
```suggestion
            const [username, email, password] = data;
```
anotherPassword не используется и не очень понимаю зачем писать лишние буквы Array.from, поясни

DPeshkoff commented on package.json:
```diff
@@ -16,17 +16,24 @@
   },
   "homepage": "https://github.com/frontend-park-mail-ru/2022_2_Base#readme",
   "dependencies": {
     "eslint": "^8.24.0",
```
Да и линтер тоже дев-депенденси
В зависимостях-зависимостях должен быть минимум, требуемый для работы, который можно ставить на тачке

DPeshkoff replied:
ping

DPeshkoff commented on public/src/components/templates.js:
```diff
@@ -0,0 +1 @@
+!function(){var n=Handlebars.template,e=Handlebars.templates=Handlebars.templates||{};e["footer.hbs"]=n({compiler:[8,">= 4.3.0"],main:function(n,e,t,o,a){return'<link rel="stylesheet" href="./src/components/Footer/Footer.css" type="text/css">\n\n<div class="footer__grid">\n    <div>\n        <img src="img/Logo.png" alt="FooterLogo">\n    </div>\n    <div class="footer__members">\n        <div cla ...
```
Этого файлика быть не должно

DPeshkoff commented on public/src/modules/ErrorMessage.js:
```diff
@@ -0,0 +1,49 @@
+     * @param {object} event событие фокусирования на элементе
+     */
+    deleteErrorMessage = (nameId) => {
+        if (document.getElementById(nameId + 'Error') !== null) {
```
Можно без !== null

DPeshkoff commented on public/src/modules/validation.js:
```diff
@@ -58,21 +28,26 @@ export default class Validation {
-        return ((password === anotherPassword) && (this.validateEMail(email).status) &&
-            (this.validatePassword(password).status));
+    checkEmptyField = (data) => {
+        if (data === "") {
```
Можно проверять длину на 0 или просто !!

DPeshkoff commented on public/src/pages/LoginPage/LoginPage.js:
```diff
@@ -45,88 +46,98 @@ export default class LoginPage extends BasePage {
         };
 
-        form.addEventListener('focusout', realTimeCheckHandler);
+        form.addEventListener('focusin', async (event) => {errorMessage.deleteErrorMessage(event.target.name);});
```
А как мы этот обработчик снимем?)

DPeshkoff commented on public/src/pages/MainPage/MainPage.js:
```diff
@@ -50,7 +50,8 @@ export default class MainPage extends BasePage {
-        const [status, outD] = await r.makeGetRequest('api/v1/').catch((err) => console.log(err));
+        const [status, outD] = await r.makeGetRequest('api/v1/products')
+            .catch((err) => console.log(err));
 
```
На сейчас ок, но в будущем надо бы сделать плашечку для юзера, чтобы он видел, что "что-то пошло не так"

DPeshkoff commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,109 @@ export default class RegisterPage extends BasePage {
+                console.log('no auth: ', status);
+                break;
+            default:
+                document.getElementById('serverErrorMessage') === null ?
```
Можно без null

DPeshkoff commented on public/src/pages/templates.js:
```diff
@@ -0,0 +1 @@
+!function(){var n=Handlebars.template,a=Handlebars.templates=Handlebars.templates||{};a["LoginPage.hbs"]=n({compiler:[8,">= 4.3.0"],main:function(n,a,i,t,s){return'<link rel="stylesheet" href="./src/pages/LoginPage/LoginPage.css" type="text/css">\n\n<div class="page">\n    <header id="header"></header>\n\n    <main>\n        <div class="paint-background"></div>\n        <div class="content">\n    ...
```
Этого файлика тоже быть не должно

DPeshkoff commented:
А еще - давай пофиксим так, чтобы проходил CI (там в шаблонах беда)

Liza1040 committed:
Delete templates.js

Liza1040 committed:
Delete templates.js

Liza1040 committed:
Поправлены ошибки, отловленные  linter

Liza1040 committed:
Merge branch 'Base-2.5_fix_validation' of https://github.com/frontend-park-mail-ru/2022_2_Base into Base-2.5_fix_validation

Liza1040 committed:
Проверка линтера

Liza1040 committed:
Проверка linter

Liza1040 committed:
Проверка linter

DPeshkoff commented on public/src/modules/validation.js:
```diff
@@ -58,21 +28,26 @@ export default class Validation {
-            (this.validatePassword(password).status));
+    checkEmptyField = (data) => {
+        if (data.length === 0) {
+            return {status: false, message: 'Поле обязательно должно быть заполнено'};
```
Вроде обсуждали, что унести в константы надо?

DPeshkoff commented on public/src/pages/LoginPage/LoginPage.js:
```diff
@@ -5,6 +5,7 @@ import FormComponent from '../../components/Form/Form.js';
 import FooterComponent from '../../components/Footer/Footer.js';
 import Req from '../../modules/ajax.js';
 import Val from '../../modules/validation.js';
+import ErrorMes from '../../modules/ErrorMessage.js';
```
```suggestion
import ErrorMessage from '../../modules/ErrorMessage.js';
```
так читабельнее)

DPeshkoff commented on public/src/pages/RegisterPage/RegisterPage.js:
```diff
@@ -45,96 +46,114 @@ export default class RegisterPage extends BasePage {
         };
 
-        form.addEventListener('focusout', realTimeCheckHandler);
+        form.addEventListener('focusin', async (event) => {
```
анонимные обработчики - плохо

Liza1040 replied:
Мы хотели сначала смерджить, а потом удалить его. Удаление в ветке у Влада

Liza1040 committed:
Сообщения об ошибке вынесено в константу.  Линтер перенесен в devDependencies.

Liza1040 committed:
Merge branch 'main' into Base-2.5_fix_validation

tUnknownLegend committed:
переписана validate

так же validation и errorMessage импортируются как синглтоны, пофикшены ошибки и линтеры

tUnknownLegend committed:
убрал auth: true в конфиге

tUnknownLegend committed:
исрпавлено извдечение данных из data при отправке запроса

tUnknownLegend committed:
исрпавил парсинг data v2

tUnknownLegend committed:
испавление удаления EventListener

Liza1040 committed:
Поправлен JSDoc.
